README.md
# Qstit
Qstit is a QT based Free Software for live projection and manipulation of electronic subtitles (also called "surtitles", "supertitles", "soft-titles"...). **Qstit** is composed of a **display** (the textframe with titles to be projected), a **control bar** (with the buttons) and a **grid** (the list of titles), all of which can be moved around on the **fullscreen background** depending on the desired set-up.

This project is distributed under GNU General Public License, version 3 or later --- see the file COPYING for details. This project was originally developed by Nova Cinema, Brussels - [subtitles.nova-cinema.org](http://subtitles.nova-cinema.org). This fork is currently maintained by Naatak Org - [naatak.com](https://naatak.com). Icons used in this project are provided by Nova & Axialis Team (with some modifications) - [www.axialis.com/free/icons](https://www.axialis.com/free/icons)

## HowTo
When a projector is plugged to the computer, it is considered as a second monitor. If the computer is set to have the same image on both screens, you will need to hide the controls from the public, the title display is the only part they need to see. Use a cover (a piece of cardboard in front of the projector can do the trick), aim the projector to hit the screen with only the top part of the image (where the display will be placed), or select only this part of the signal using a video-mixer. You can also configure your computer display to have two separate screens. In that case you need to have the controls on your screen and only the display and the fullscreen background on the projector. This can already work but it can be a bit tricky as **Qstit** doesn't fully support the detection of multiple screens yet.

### Supported file formats
**Qstit** can read simple text (.txt) with up to four lines per title. **Qstit** provides support for UTF-8 encoding which makes it possible to display all sorts of characters in all languages.

For `.txt` file type the structure (detailed below) needs to be respected carefully otherwise **Qstit** won't be able to open them correctly. If errors are found when opening the file, a message is displayed. The errors encountered are listed in the 'File info' window and also saved in a log file in the same folder as the title file (the **log file** is also created if titles are flagged or if modifications have been made using the 'Edit' functionality).

`.txt` can include the following HTML tags to change the style of the text in between them:
```
<i> _italic_ </i>
```
```
<b> **bold** </b>
```
```
<u> underline </u>
```

A tag can start on one line and end on another, wherever in the file. Whole sections can easily be turned into italic, for instance. The tags themselves are not displayed but appear in the grid. Everything written between an opening (<) and a closing (>) bracket (or a line break) will be considered as a tag and won't be displayed. An entire line between brackets will still appear in the grid like a normal title but won't appear on the display. Brackets can thus be used to put comments and cues in the file.

#### Text (.txt) files

Rules
- One line in the file = one row in the grid = one title.
- One title can be composed of several lines (up to four, usually two), they are separated by a backslash (\).
- Different titles can be composed of different numbers of lines.
· Titles are only separated by a line break.
- **Qstit** automatically creates "empty" lines between titles so that nothing is displayed between two titles when there are no dialogues. There is no need to have empty lines in the file, they would be considered as one title, even if there isn't any text. There would therefore be three consecutive "empty lines" when browsing through the file.

##### Format

```
Title 1 Line 1\Title 1 Line 2
Title 2 Line 1
Title 3, Line 1\Title 3, Line 2\Title 3, Line 3
Title 4, Line 1\Title 4, Line 2
...
```

Example: Shows the first title is italized, subsequently the second title is bolder but has 2 more subsequent lines.

```
<i>Welcome to Naatak's Gujarati Mela 2018</i>
<b>FITTING ROOMS</b>\ Author: Susan Cinoman,Translation:Sameer Pandya\ Director: Natraj Kumar
MADHURI: Hmm... If I buy this, my husband will kill me.
LEENA: You are kidding. Right?
MADHURI: Ofcourse I don't mean that literally but he will not like it.
LEENA: Well, who cares about that! It is not like he makes you happy anyways.
MADHURI: But can anyone really make anyone else happy?
LEENA: What do you mean?
MADHURI: One needs to keep oneself happy. Right?
LEENA: Good then, you found a good excuse to buy this dress.
...
```

### Controls
**Qstit** can be controlled with the mouse and / or keyboard. The controls and keyboard shortcuts are detailed below.
It is possible to fully control **Qstit** with just a numeric keypad (or "numpad"). An external keypad can therefore be used as a remote control.

#### Menu

![](src/Imag/QSTit_sett_enab.png) Settings

Show 'Settings' window

\[Ctrl+S\]

![](src/Imag/QSTit_file.png) Open

Open title file (Supported formats: .txt)

\[Ctrl+O\]

![](src/Imag/QSTit_help_enab.png) Help

Show this window

\[F1\]

![](src/Imag/QSTit_exit.png) Quit

Quit the program

\[Ctrl+Q\]

#### Titles control

![](src/Imag/QSTit_upup.png)

Previous title / empty line

\[↑\] or \[Numpad 8\]

![](src/Imag/QSTit_down.png)

Next title / empty line With .txt files, also works with:

\[↓\] or \[Numpad 2\]
\[Space bar\] or \[Numpad 0\]

![](src/Imag/QSTit_play.png)

Start / Pause timer

\[Space bar\] or \[Numpad 0\]

#### Display control

![](src/Imag/QSTit_roll_open.png) / ![](src/Imag/QSTit_roll_clos.png)

Hide / Show title display
(When on pause, restarting makes the display appear)

\[X\] or \[Numpad .\]

Move display up

\[Page Up\] or \[Numpad 9\]

Move display down

\[Page Down\] or \[Numpad 3\]

Move display left

\[Home\] or \[Numpad 7\]

Move display right

\[End\] or \[Numpad 1\]

→ use \[Shift\] for faster movements

It is also possible to move the display with the mouse

\[Click and drag\]

#### Grid

![](src/Imag/QSTit_info_enab.png) File info

Show 'File info' window

\[Ctrl+I\]

![](src/Imag/QSTit_save_enab.png) Save as...

Save modified file as...

![](src/Imag/QSTit_edit_enab.png) Edit

Edit selected text / timecode in the grid
(Edits are indicated in the grid and listed in the log file)

\[Ctrl+E\]

![](src/Imag/QSTit_undo_enab.png) Undo

Undo changes (in reverse chronology)

\[Ctrl+Z\]

![](src/Imag/QSTit_goto_enab.png) Go to...

Go to a chosen title

\[Ctrl+G\]

![](src/Imag/QSTit_find_enab.png) Find...

Find text in the file

\[Ctrl+F\]

![](src/Imag/QSTit_clea.png) Clear

Clear display and reset file at beginning

\[Ctrl+C\]

![](src/Imag/QSTit_link_none.png) Bind/Unbind

(Un)bind menu bar and grid for custom positioning

![](src/Imag/QSTit_flag.png) (Leftmost cell)

Flag current row to indicate a cue, a mistake, etc.
(Flags are saved in the log file)

\[!\] or \[Numpad /\]

#### Advanced Settings
Various display and control settings can be adjusted in the 'Settings' window. The display can be adapted line per line or all lines together. Note that you can make a line "**invisible**" using the advanced settings. The full text will still appear in the grid for your control but it won't be displayed. It can be useful if the titles are one language on Line 1 and another on Line 2, for instance, and you only need to project one of the two but want to keep the translation in the file.

Settings can be saved for later use with the current file or as general settings for all files. **Settings saved for a specific file** will be named like it (with a .cfg extension) and placed in the same folder as the subtitle file. **General settings** are saved in a file named "qstit.cfg" which will be created in your "home" or "user" folder, depending on your system (in a ".qstit" subfolder). There are also **default settings** built in the program that can be restored to replace your configurations, if needed. If you don't need the settings anymore, you can just delete the .cfg files.

When opening a new file, you can choose which settings to use. Once a file is opened, you can always load preset settings using the designated functions in the 'Save / Load' tab. It is recommended to test your settings while displaying one of the the widest lines, which are listed in the title bar and in the 'File info' window.



debian/changelog
qstit (3.0.0-1) stable; urgency=low

  * Update the fork with recent fork

 -- Harshavardhana <harsha@naatak.com>  Tue, 31 Dec 2018 14:28:05 +0200

qstit (2.9.6-1) stable; urgency=low

  * Fixes bug in the previous version when resuming the timer.

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Tue, 24 September 2013 14:28:05 +0200

qstit (2.9.5-1) stable; urgency=low

  * Changes in the timer for more accurate timing (particularly notable on Windows)
  * Correction of the positioning based on saved settings for files with less than 4 lines

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Tue, 17 September 2013 14:28:05 +0200

qstit (2.9.4-1) stable; urgency=low

  * Small correction for the subtitle frame size calculation on Windows/Linux
  * No other changes, same as 2.9.3 for Mac

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Tue, 10 September 2013 10:19:12 +0200

qstit (2.9.3-1) stable; urgency=low

  * Small correction for the subtitle frame size calculation on Mac
  * No other changes, same as 2.9.2 for Linux or Windows

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Tue, 15 August 2013 22:19:12 +0200

qstit (2.9.2-1) stable; urgency=low

  * Corrected subtitle frame size calculation for Mac
  * Help window resize
  * Disable unexisting lines in the settings
  * Clarify visibility setting for the lines
  * Added test screen for calibration and debug
  * Various fixes

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Tue, 30 July 2013 18:39:25 +0200

qstit (2.9.1-1) stable; urgency=low

  * Help file modified
  * Date format for 'file info' window

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Tue, 09 July 2013 11:39:25 +0200

qstit (2.9.0-1) stable; urgency=low

  * Settings system rewritten
  * various corrections

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Sun, 16 June 2013 11:45:40 +0200

qstit (2.8.6-1) stable; urgency=low

  * Updated Debian package info
  * Help file modified

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Mon, 13 May 2013 11:45:40 +0200

qstit (2.8.5-1) stable; urgency=low

  * Small style changes

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Sat, 04 May 2013 16:00:18 +0200

qstit (2.8.4-1) stable; urgency=low

  * Styles

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Sat, 04 May 2013 15:59:28 +0200

qstit (2.8.3-1) stable; urgency=low

  * help → png
  * Grid and menu style sheets

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Fri, 19 Apr 2013 16:10:15 +0200

qstit (2.8.2-1) stable; urgency=low

  * Config path for Linux
  * Grid and menu style sheets
  * Jump base → 0.2

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Wed, 17 Apr 2013 13:18:31 +0200

qstit (2.8.1-1) stable; urgency=low

  * Initial debian package release

 -- Laurent Tenzer (Qstit) <subtitles@nova-cinema.org>  Sun, 14 Apr 2013 21:41:52 +0200


src/QSTit_diaGrid.cpp
/*
    This file is part of Qstit - http://subtitles.nova-cinema.org
    Feedback, comments and questions are welcome: subtitles@nova-cinema.org
   ════════════════════════════════════════════════════════════════════════════════════
    Qstit is developed by Nova Cinema, Brussels - http://www.nova-cinema.org
    Conception: Laurent Tenzer
    Programming: Georges Piedboeuf-Boen - georges.pi.bo@gmail.com
    Icons credit: Nova & Axialis Team - http://www.axialis.com/free/icons
   ════════════════════════════════════════════════════════════════════════════════════
    Qstit is a free software: you can redistribute it and/or modify it under the terms
    of the GNU General Public License as published by the Free Software Foundation,
    either version 3 of the License, or (at your option) any later version.
    If you do, we'd like to hear about it.

    Qstit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
    without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
    PARTICULAR PURPOSE. See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along with Qstit.
    If not, see http://www.gnu.org/licenses
   ════════════════════════════════════════════════════════════════════════════════════
*/

#include "QSTit_diaGrid.h"
#include "QSTit_diaSkin.h"
#include "QSTit_diaButt.h"

diaGridSkin::diaGridSkin(QWidget *parent,menuSkin *menu,QString pTitl,int pWidt,int pHeig) : QFrame(parent)
{
    parWin=parent;
    parMen=menu;
    iPr=QApplication::desktop()->width()-1;
    iPb=QApplication::desktop()->height()-1;
    iPm=QApplication::desktop()->height()/2;
    sPf=qApp->font().family();
    bLink=true;
    iW=pWidt;
    iH=pHeig;
    iX=fCalcPosX(iW);
    iY=fCalcPosY(iH,parMen);
    sStylBord="border:none;background-color:transparent;";

    this->setStyleSheet(fStylFram());
    this->setMinimumSize(QSize(500,iH));
    this->setMaximumSize(QSize(iW,iH+160));
    this->setContentsMargins(0,0,0,0);
    this->setCursor(Qt::ArrowCursor);
    this->setGeometry(iX,iY,iW,iH);
    this->setHidden(true);

    labTitl=new QLabel(pTitl,this);
    labTitl->setStyleSheet(fStylTitl());
    labTitl->setGeometry(0,0,iW,30);

    fraSizT=new QFrame(this);
    fraSizT->setGeometry(0,0,iW,5);
    fraSizT->raise();
    fraSizT->setCursor(QPixmap(":/Imag/QSTit_size_vert.png"));
    fraSizT->setFrameShape(QFrame::NoFrame);
    fraSizT->setStyleSheet(sStylBord);
    fraSizL=new QFrame(this);
    fraSizL->setGeometry(0,0,5,iH);
    fraSizL->raise();
    fraSizL->setCursor(QPixmap(":/Imag/QSTit_size.png"));
    fraSizL->setFrameShape(QFrame::NoFrame);
    fraSizL->setStyleSheet(sStylBord);
    fraSizR=new QFrame(this);
    fraSizR->setGeometry(iW-5,0,5,iH);
    fraSizR->setCursor(QPixmap(":/Imag/QSTit_size.png"));
    fraSizR->setFrameShape(QFrame::NoFrame);
    fraSizR->setWindowOpacity(0);
    fraSizR->setStyleSheet(sStylBord);
    fraSizB=new QFrame(this);
    fraSizB->setGeometry(0,iH-5,iW,5);
    fraSizB->raise();
    fraSizB->setCursor(QPixmap(":/Imag/QSTit_size_vert.png"));
    fraSizB->setFrameShape(QFrame::NoFrame);
    fraSizB->setStyleSheet(sStylBord);
}
void diaGridSkin::fRePosi()
{
    iX=fCalcPosX(iW);
    if (bLink) iY=fCalcPosY(iH,parMen);
    bool bLinkSave=bLink;
    bLink=false;                                        // supress link to avoid menu repositioning
    this->move(iX,iY);
    bLink=bLinkSave;
}
void diaGridSkin::fGridLink()
{
    bool bL=!bLink;                                     // must be so due to fMenuMove that may change bLink
    parMen->bLink=bL;
    if (bL) fMenuMove();
    bLink=bL;
}
void diaGridSkin::fMenuMove()
{
    parMen->iX=iX+((iW-parMen->frameGeometry().width())/2);
    if(this->frameGeometry().y()+(iH/2)>iPm) parMen->iY=iY+this->frameGeometry().height();
    else parMen->iY=iY-parMen->frameGeometry().height();
    parMen->move(parMen->iX,parMen->iY);
}
void diaGridSkin::fGridSizeInit(int pW,int pRowsMaxi,int pConf)
{
    if (pRowsMaxi<16) this->setMaximumHeight(34+(pRowsMaxi*18));
    else this->setMaximumHeight(34+(16*18));

    if (pW>=this->minimumWidth())iW=pW;
    else iW=this->minimumWidth();
    if (pConf<=1 || iX+iW>iPr) iX=(iPr-iW)/2;
    this->setGeometry(iX,iY,iW,iH);
    int iw=iW-81;
    QList<diaGridC *> widg=this->findChildren<diaGridC *>();
    widg[0]->resize(iw,iH-33);
    widg[0]->setColumnWidth(3,iWid3);dPou3=(double)iWid3/iw;
    widg[0]->setColumnWidth(4,iWid4);dPou4=(double)iWid4/iw;
    widg[0]->setColumnWidth(5,iWid5);dPou5=(double)iWid5/iw;
    widg[0]->setColumnWidth(6,iWid6);dPou6=(double)iWid6/iw;
    fGridSizeBord();
    QList<butToolC *> butt=this->findChildren<butToolC *>();
    butt[0]->move(iW-69,butt[0]->frameGeometry().y());
    butt[1]->move(iW-39,butt[1]->frameGeometry().y());
    butt[2]->move(iW-69,butt[2]->frameGeometry().y());
    butt[3]->move(iW-39,butt[3]->frameGeometry().y());
    butt[4]->move(iW-69,butt[4]->frameGeometry().y());
    butt[5]->move(iW-39,butt[5]->frameGeometry().y());
    butt[6]->move(iW-69,butt[6]->frameGeometry().y());
    butt[7]->move(iW-39,butt[7]->frameGeometry().y());
    if (pConf<=1 && bLink) fMenuMove();
}
void diaGridSkin::fGridSizeVert(bool pB,int pD)
{
    int jH;
    if (pB) jH=iH-pD;
    else jH=iH+pD;
    if (jH>=this->minimumHeight() && jH<=this->maximumHeight())
    {
        iH=jH;
        this->resize(iW,iH);
        if (bLink)
        {
            iY=fCalcPosY(iH,parMen);
            bool bLinkSave=bLink;
            bLink=false;                                // supress link to avoid menu repositioning
            this->move(iX,iY);
            bLink=bLinkSave;
        }
        else
        {
            if (iY>iPb-iH) iY=iPb-iH;
            this->move(iX,iY);
        }
        QList<diaGridC *> widg=this->findChildren<diaGridC *>();
        widg[0]->resize(widg[0]->frameGeometry().width(),iH-33);
        fGridSizeBord();
        if (bLink) fMenuMove();
        emit sMoved();
    }
}
void diaGridSkin::fGridSizeHori(bool pL,int pW)
{
    int jX,jW;
    if (pL) {jX=pW;jW=this->frameGeometry().x()-pW+this->frameGeometry().width();}
    else {jX=this->frameGeometry().x();jW=pW-jX;}
    if (jW>=this->minimumWidth() && jW<=this->maximumWidth())
    {
        if (pL) {iX=jX;this->move(iX,iY);}
        iW=jW;
        this->resize(iW,iH);
        int iw=iW-81;
        QList<diaGridC *> widg=this->findChildren<diaGridC *>();
        widg[0]->resize(iw,widg[0]->frameGeometry().height());
        if (nColo>0) widg[0]->setColumnWidth(3,(int)iw*dPou3);
        if (nColo>1) widg[0]->setColumnWidth(4,(int)iw*dPou4);
        if (nColo>2) widg[0]->setColumnWidth(5,(int)iw*dPou5);
        if (nColo>3) widg[0]->setColumnWidth(6,(int)iw*dPou6);
        fGridSizeBord();
        QList<butToolC *> butt=this->findChildren<butToolC *>();
        butt[0]->move(iW-69,butt[0]->frameGeometry().y());
        butt[1]->move(iW-39,butt[1]->frameGeometry().y());
        butt[2]->move(iW-69,butt[2]->frameGeometry().y());
        butt[3]->move(iW-39,butt[3]->frameGeometry().y());
        butt[4]->move(iW-69,butt[4]->frameGeometry().y());
        butt[5]->move(iW-39,butt[5]->frameGeometry().y());
        butt[6]->move(iW-69,butt[6]->frameGeometry().y());
        butt[7]->move(iW-39,butt[7]->frameGeometry().y());
        if (bLink) fMenuMove();
        emit sMoved();
    }
}
void diaGridSkin::fGridSizeBord()
{
    labTitl->resize(iW,30);
    fraSizT->resize(iW,5);
    fraSizL->resize(5,iH);
    fraSizB->setGeometry(0,iH-5,iW,5);
    fraSizR->setGeometry(iW-5,0,5,iH);
}
void diaGridSkin::mousePressEvent(QMouseEvent *e)
{
    bSizT=(e->y()>=0 && e->y()<=5)?true:false;
    bSizL=(e->y()>=0 && e->x()<=5)?true:false;
    bSizB=(e->y()>=iH-5 && e->y()<=iH)?true:false;
    bSizR=(e->y()>=0 && e->x()>=iW-5)?true:false;
    iCx=e->globalX();
    iCy=e->globalY();
}
void diaGridSkin::mouseMoveEvent(QMouseEvent *e)
{
    if (bSizT) {fGridSizeVert(false,iCy-e->globalY());iCx=e->globalX();iCy=e->globalY();}
    else if (bSizL) {fGridSizeHori(true,e->globalX());iCx=e->globalX();iCy=e->globalY();}
    else if (bSizB) {fGridSizeVert(true,iCy-e->globalY());iCx=e->globalX();iCy=e->globalY();}
    else if (bSizR) {fGridSizeHori(false,e->globalX());iCx=e->globalX();iCy=e->globalY();}
    else
    {
        int jX=iX+e->globalX()-iCx;
        if (jX<1) jX=1;
        if (jX>iPr-iW) jX=iPr-iW;
        if (bLink && parMen->iX<1) jX=iX+1;
        if (bLink && parMen->iX>iPr-parMen->iW) jX=iX-1;
        iX=jX;
        int jY=iY+e->globalY()-iCy;
        if (jY<1) jY=1;
        if (bLink && jY<41) jY=41;
        if (jY>iPb-iH) jY=iPb-iH;
        if (bLink && jY>iPb-iH-40) jY=iPb-iH-40;
        iY=jY;
        this->move(iX,iY);
        iCx=e->globalX();
        iCy=e->globalY();
        emit sMoved();
    }
}
void diaGridSkin::moveEvent(QMoveEvent*)
{
    iX=this->frameGeometry().x();
    iY=this->frameGeometry().y();
    if (bLink) fMenuMove();
}
void diaGridSkin::fWorkLang(QStringList pCode,QStringList pText)
{
    gLangCode=pCode;
    gLangText=pText;
}
QString diaGridSkin::fL(QString sI)
{
    int i=gLangCode.indexOf(sI,0);
    if (i>-1) sI=gLangText.at(i);
    return sI;
}

//-------------------------------------------------------------------------------------------------

diaGridC::diaGridC(diaGridSkin *parent,QWidget *wind,QFrame *menu) : QTableWidget(parent)
{
    parEnt=parent;
    parWin=wind;
    parMen=menu;
    gFindText="";
    gFindFrom=0;
}
void diaGridC::keyPressEvent(QKeyEvent *e)
{
    if (e->modifiers()==Qt::NoModifier)
    {
        switch (e->key())
        {
            case Qt::Key_Exclam:    {fGridMarkRows();break;}
            case Qt::Key_F3:        {if (!gFindText.isEmpty()) {fGridFindLoop(gFindText,gFindFrom);}};break;
            default: {QApplication::sendEvent(parWin,new QKeyEvent(QEvent::KeyPress,e->key(),e->modifiers(),"",false,1));}
        }
    }
    else if (e->modifiers()==Qt::ControlModifier)
    {
        switch (e->key())
        {
            case Qt::Key_G:         {fGridGoto();break;}
            case Qt::Key_F:         {fGridFind();break;}
            default: {QApplication::sendEvent(parWin,new QKeyEvent(QEvent::KeyPress,e->key(),e->modifiers(),"",false,1));}
        }
    }
    else
    {
        QApplication::sendEvent(parWin,new QKeyEvent(QEvent::KeyPress,e->key(),e->modifiers(),"",false,1));
    }
}
void diaGridC::fGridGoto()
{
    QString sPf=qApp->font().family();

    int iLineWidt=50;
    int iLineHeig=20;
    int iWidt=iLineWidt+95;
    int iHeig=65;
    int iLefx=parEnt->frameGeometry().x()+parEnt->frameGeometry().width()-iWidt-79;
    int iTopy=parEnt->frameGeometry().y()+2;

    diaGoto=new QDialog(parEnt);
    diaGoto->setWindowFlags(Qt::Window|Qt::FramelessWindowHint);
    diaGoto->setModal(true);
    diaGoto->setCursor(Qt::ArrowCursor);
    diaGoto->setMinimumSize(iWidt,iHeig);
    diaGoto->setMaximumSize(iWidt,iHeig);
    diaGoto->move(iLefx,iTopy);

    QString sStyl="QDialog {font:normal 11px "+sPf+";background-color:#33ff33;color:#000000;border:none}";
    sStyl+="QLineEdit {background-color:#ffffff;color:#000000;border:1px solid #202020;font:11px normal courier new;}";
    sStyl+="QLabel {background-color:#33ff33;color:#202020;border:none;text-align:left;}";
    sStyl+="QPushButton {font:normal 11px "+sPf+";min-height:20px;max-height:20px;min-width:70px;max-width:90px;background-color:#202020;color:#ffffff;border:1px solid #555555;border-radius:3px;text-align:left;padding-left:2px;}";
    sStyl+="QPushButton:hover {border-color:#cccccc;}";


    diaGoto->setStyleSheet(sStyl);

    ediGoto=new QLineEdit(diaGoto);
    ediGoto->setEchoMode(QLineEdit::Normal);
    ediGoto->setText("");
    ediGoto->setGeometry(10,10,iLineWidt,iLineHeig);
    ediGoto->setFocus();
    connect(ediGoto,SIGNAL(returnPressed()),this,SLOT(fGridGotoLine()));

    QLabel *labGoto=new QLabel(QString("1 → %1").arg(rowCount()-2),diaGoto);
    labGoto->move(10,35);

    QPushButton *butGotoSave=new QPushButton(QPixmap(":Imag/QSTit_okok.png"),fL("butGridGotoSave"),diaGoto);
    butGotoSave->setIconSize(QSize(16,16));
    butGotoSave->resize(70,20);
    butGotoSave->move(iLineWidt+15,10);
    connect(butGotoSave,SIGNAL(clicked()),this,SLOT(fGridGotoLine()));

    QPushButton *butGotoCanc=new QPushButton(QPixmap(":Imag/QSTit_canc.png"),fL("butGridGotoCanc"),diaGoto);
    butGotoCanc->setIconSize(QSize(16,16));
    butGotoCanc->resize(70,20);
    butGotoCanc->move(iLineWidt+15,35);
    connect(butGotoCanc,SIGNAL(clicked()),this,SLOT(fGridGotoCanc()));

    gGridGoto=false;
    diaGoto->exec();

    if (gGridGoto)
    {
        QString sR=ediGoto->text();
        int iR=sR.toInt();
        setCurrentCell(iR,3,QItemSelectionModel::Rows);
        scrollTo(currentIndex(),QAbstractItemView::PositionAtCenter);
    }

    activateWindow();
    setFocus();
}
void diaGridC::fGridGotoCanc()
{
    gGridGoto=false;
    diaGoto->done(0);
}
void diaGridC::fGridGotoLine()
{
    QString sR=ediGoto->text();
    int iR=sR.toInt();
    int iM=rowCount()-1;
    if (iR>0&&iR<iM)
    {
        gGridGoto=true;
        diaGoto->done(0);
    }
}
void diaGridC::fGridFind()
{
    QString sPf=qApp->font().family();

    int iLineWidt=100;
    int iLineHeig=20;
    int iWidt=iLineWidt+95;
    int iHeig=65;
    int iLefx=parEnt->frameGeometry().x()+parEnt->frameGeometry().width()-iWidt-79;
    int iTopy=parEnt->frameGeometry().y()+2;

    diaFind=new QDialog(parEnt);
    diaFind->setWindowFlags(Qt::Window|Qt::FramelessWindowHint);
    diaFind->setModal(true);
    diaFind->setCursor(Qt::ArrowCursor);
    diaFind->setMinimumSize(iWidt,iHeig);
    diaFind->setMaximumSize(iWidt,iHeig);
    diaFind->move(iLefx,iTopy);

    QString sStyl="QDialog {font:normal 11px "+sPf+";background-color:#33ff33;color:#000000;border:none}";
    sStyl+="QLineEdit {background-color:#ffffff;color:#000000;border:1px solid #202020;}";
    sStyl+="QLabel {background-color:#33ff33;color:#202020;border:none;text-align:left;}";
    sStyl+="QPushButton {font:normal 11px "+sPf+";min-height:20px;max-height:20px;min-width:70px;max-width:90px;background-color:#202020;color:#ffffff;border:1px solid #555555;border-radius:3px;text-align:left;padding-left:2px;}";
    sStyl+="QPushButton:hover {border-color:#cccccc;}";
    diaFind->setStyleSheet(sStyl);

    ediFind=new QLineEdit(diaFind);
    ediFind->setEchoMode(QLineEdit::Normal);
    ediFind->setText("");
    ediFind->setGeometry(10,10,iLineWidt,iLineHeig);
    ediFind->setFocus();
    connect(ediFind,SIGNAL(returnPressed()),this,SLOT(fGridFindWord()));

    QLabel *labFind=new QLabel(fL("labGridFind"),diaFind);
    labFind->move(10,35);

    QPushButton *butFindSave=new QPushButton(QPixmap(":Imag/QSTit_okok.png"),fL("butGridFindSave"),diaFind);
    butFindSave->setIconSize(QSize(16,16));
    butFindSave->resize(70,20);
    butFindSave->move(iLineWidt+15,10);
    connect(butFindSave,SIGNAL(clicked()),this,SLOT(fGridFindWord()));

    QPushButton *butFindCanc=new QPushButton(QPixmap(":Imag/QSTit_canc.png"),fL("butGridFindCanc"),diaFind);
    butFindCanc->setIconSize(QSize(16,16));
    butFindCanc->resize(70,20);
    butFindCanc->move(iLineWidt+15,35);
    connect(butFindCanc,SIGNAL(clicked()),this,SLOT(fGridFindCanc()));

    gGridFind=false;
    diaFind->exec();

    if (gGridFind)
    {
        QString sT=ediFind->text();
        if (!sT.isEmpty())
        {
            gFindText=sT;
            gFindFrom=0;
            fGridFindLoop(gFindText,gFindFrom);
        }
    }

    activateWindow();
    setFocus();
}
void diaGridC::fGridFindCanc()
{
    gGridFind=false;
    diaFind->done(0);
}
void diaGridC::fGridFindWord()
{
    gGridFind=true;
    diaFind->done(0);
}
void diaGridC::fGridFindLoop(QString pText,int pFrom)
{
    QString sFind=pText.toLower();
    int iR;
    int iRM=rowCount();
    int iC;
    int iCM=columnCount();
    QTableWidgetItem *it;
    QString st;

    for (iR=pFrom; iR<iRM; iR++)
    {
        for (iC=1; iC<iCM; iC++)
        {
            if (columnWidth(iC)>0)
            {
                it=item(iR,iC);
                st=it->text().toLower();
                if (st.indexOf(sFind)!=-1)
                {
                    setCurrentCell(iR,iC,QItemSelectionModel::Rows);
                    gGridFind=true;
                    gFindFrom=iR+1;
                    diaFind->done(0);
                    return;
                }
            }
        }
    }
    messSkin cM;
    cM.fMess(this,parMen,fL("butGridFindEnd"),fL("butOkOk"),"");
}
void diaGridC::fGridMarkRows()
{
    int iR=currentRow();

    if (iR==0||iR==rowCount()-1) return;

    QTableWidgetItem *wi=item(iR,0);

    if (wi->text()=="E") return;

    QIcon ic=wi->icon();
    if (!ic.isNull())
    {
        QTableWidgetItem *wi=new QTableWidgetItem("",0);
        wi->setFlags(Qt::NoItemFlags|Qt::ItemIsSelectable|Qt::ItemIsEnabled);
        setItem(iR,0,wi);
    }
    else
    {
        QTableWidgetItem *wi=item(iR,0);
        wi->setIcon(QPixmap(":Imag/QSTit_flag.png"));
        wi->setText("I");
    }
}
void diaGridC::fWorkLang(QStringList pCode,QStringList pText)
{
    gLangCode=pCode;
    gLangText=pText;
}
QString diaGridC::fL(QString sI)
{
    int i=gLangCode.indexOf(sI,0);
    if (i>-1) sI=gLangText.at(i);
    return sI;
}
